cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java

```diff
@@ -0,0 +1,259 @@
+            return authentication;
+        } else if (mandatory) {
+            redirectToCas(request, response);
+            return Authentication.SEND_CONTINUE;
+        }
+        return Authentication.UNAUTHENTICATED;
+    }
+
+    @Override
+    public boolean secureResponse(
+            final ServletRequest request,
+            final ServletResponse response,
+            final boolean mandatory,
+            final Authentication.User user) throws ServerAuthException {
+        return true;
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        if (ticketValidator == null) {
+            throw new RuntimeException("TicketValidator cannot be null");
+        }
+        if (serverNames == null) {
+            throw new RuntimeException("ServerNames cannot be null");
+        }
+    }
+
+    protected void clearCachedAuthentication(final String ticket) {
+        final WeakReference<HttpSession> sessionRef = sessionMap.remove(ticket);
+        if (sessionRef != null && sessionRef.get() != null) {
+            sessionRef.get().removeAttribute(CACHED_AUTHN_ATTRIBUTE);
+        }
+    }
+
+    private void cacheAuthentication(final HttpServletRequest request, final CasAuthentication authentication) {
+        final HttpSession session = request.getSession(true);
+        if (session != null) {
+            session.setAttribute(CACHED_AUTHN_ATTRIBUTE, authentication);
+            sessionMap.put(authentication.getTicket(), new WeakReference<HttpSession>(session));
+        }
+    }
+
+    private CasAuthentication fetchCachedAuthentication(final HttpServletRequest request) {
+        final HttpSession session = request.getSession(false);
+        if (session != null) {
+            return (CasAuthentication) session.getAttribute(CACHED_AUTHN_ATTRIBUTE);
+        }
+        return null;
+    }
+
+    private String serviceUrl(final HttpServletRequest request, final HttpServletResponse response) {
+        return CommonUtils.constructServiceUrl(
+                request,
+                response,
+                null,
+                serverNames,
+                protocol.getServiceParameterName(),
+                protocol.getArtifactParameterName(),
+                true);
+    }
+
+    private void redirectToCas(
+            final HttpServletRequest request, final HttpServletResponse response) throws ServerAuthException {
+        try {
+            final String redirectUrl = CommonUtils.constructRedirectUrl(
+                    casServerLoginUrl, protocol.getServiceParameterName(), serviceUrl(request, response), renew, false);
+            logger.debug("Redirecting to {}", redirectUrl);
+            response.sendRedirect(redirectUrl);
+        } catch (IOException e) {
+            logger.debug("Redirect to CAS failed with error: {}", e);
+            throw new ServerAuthException("Redirect to CAS failed", e);
+        }
+    }
+
+    private String extractTicket(final HttpServletRequest request) {
```

Comment:
Good suggestion, don't even need that method now using instance field `protocol`. Fixed in c1c691d.